std/regex/internal/ir.d

```diff
@@ -543,25 +547,34 @@ class CtfeFactory(alias EngineType, Char, alias func) : GenericFactory!(EngineTy
 // A workaround for R-T enum re = regex(...)
 template defaultFactory(Char)
 {
-    @property MatcherFactory!Char defaultFactory(const ref Regex!Char re) @safe
+    @property MatcherFactory!Char defaultFactory(const ref Regex!Char re) @safe pure
     {
-        import std.regex.internal.backtracking : BacktrackingMatcher;
-        import std.regex.internal.thompson : ThompsonMatcher;
-        import std.algorithm.searching : canFind;
-        static MatcherFactory!Char backtrackingFactory;
-        static MatcherFactory!Char thompsonFactory;
-        if (re.backrefed.canFind!"a != 0")
-        {
-            if (backtrackingFactory is null)
-                backtrackingFactory = new RuntimeFactory!(BacktrackingMatcher, Char);
-            return backtrackingFactory;
-        }
-        else
-        {
-            if (thompsonFactory is null)
-                thompsonFactory = new RuntimeFactory!(ThompsonMatcher, Char);
-            return thompsonFactory;
-        }
+        auto defaultFactoryImpl = () {
+            import std.regex.internal.backtracking : BacktrackingMatcher;
+            import std.regex.internal.thompson : ThompsonMatcher;
+            import std.algorithm.searching : canFind;
+            static MatcherFactory!Char backtrackingFactory;
+            static MatcherFactory!Char thompsonFactory;
+            if (re.backrefed.canFind!"a != 0")
+            {
+                if (backtrackingFactory is null)
+                    backtrackingFactory = new RuntimeFactory!(BacktrackingMatcher, Char);
+                return backtrackingFactory;
+            }
+            else
+            {
+                if (thompsonFactory is null)
+                    thompsonFactory = new RuntimeFactory!(ThompsonMatcher, Char);
+                return thompsonFactory;
+            }
+        };
+
+        // this should be faked as pure because the static mutable variables are
+        // used to cache the created instance, like memoize
```

Comment:
In what sense is it valid or at least useful to mark as pure a function that accesses mutable static data? memoize isn't pure.